### ISSUE TITLE
pk-debconf-helper CPU 100% during updates

### ISSUE THREAD

**Issue by gorghino** [PackageKit/PackageKit#452]
OS: Pop!_OS 20.10
Kernel Boot Options:.quiet loglevel=0 systemd.show_status=false splash button.lid_init_state=open
Kernel Image Path:.../boot/vmlinuz-5.8.0-7630-generic
Everytime I update packages with Pop!_Shop, pk-debconf-helper ramps up to 100% CPU for ~10sec (installation).

```
gorgo@pop-os:~$ apt search libpackagekit*
Sorting... Done
Full Text Search... Done
libpackagekit-glib2-18/groovy,now 1.1.13-2ubuntu2 amd64 [installed,automatic]
  Library for accessing PackageKit using GLib

libpackagekit-glib2-dev/groovy 1.1.13-2ubuntu2 amd64
  Library for accessing PackageKit using GLib (development files)

libpackagekitqt5-1/groovy 1.0.2-1 amd64
  Library for accessing PackageKit using Qt5

libpackagekitqt5-dev/groovy 1.0.2-1 amd64
  Library for accessing PackageKit using Qt5 (development files)
```

Log verbose:
[packageKit_update_020221.txt](https://github.com/hughsie/PackageKit/files/5914441/packageKit_update_020221.txt)

**Comment by jondo** [comment]
I also see this after updates on my Linux Mint 21 (based on Ubuntu 22.04).
Others also reported this e.g. [here](https://askubuntu.com/questions/1318465/pk-debconf-helper-process-loads-a-single-processor-core-at-100-after-each-updat) and [here](https://ubuntuforums.org/showthread.php?t=2456247).

**Comment by Juno-ko** [comment]
This feature still exists. CPU usage jumps from 10% to 100% for like 15-30 seconds. OS: Pop!_OS 22.04 LTS

**Comment by Mikizeta** [comment]
Idk if this issue should be only for Pop!_OS users, but the same behavior happens also on Ubuntu 22.04 LTS as of today.

**Comment by Detzler** [comment]
I've started the software updates in ubuntu. Then the screens went black. I Logged via CTRL+ALT+F5 (or something) into the cli and saw that the cpu was eaten by pk-debconf-helper. only hard reset helped to exit the black screen.

**Comment by MissingUser** [comment]
Same bothers to me still today.
Ubuntu 22.04.3 LTS x86_64
Kernel 6.2.0.39-generic
Gnome Shell 42.9
And no info found about the task of this program.

**Comment by GitTom-HH** [comment]
I have been observing this on Ubuntu for a long time. Currently on 23.10. Here, by the way, it is only 25% CPU utilization. With a four-core processor, one core is probably fully utilized.
My avoidance strategy is to update via the command line. Nala works very well here. Still kind of unattractive.

OS: Ubuntu 23.10 x86_64 
Host: HP ENVY Notebook 
Kernel: 6.5.0-26-generic 
DE: GNOME 45.2 
CPU: Intel i7-7500U (4) @ 3.500GHz 
GPU: Intel HD Graphics 620 
Memory: 5522MiB / 15868MiB

**Comment by ximion** [comment]
In all of these cases, does the daemon still exit after a minute?

**Comment by GitTom-HH** [comment]
I had a very small update today that took about a minute. Afterwards the daemon was still there and there was no pk-debconf-helper with 25% CPU consumption. As I said, it was a very small update. 
I'll report on how it behaves when there is more to process.

**Comment by sidt4** [comment]
@GitTom-HH: Are you using `GNOME Software` or `pkcon update` in Ubuntu 23.10 ?

 Can you list the steps to reproduce this issue ?

**Comment by GitTom-HH** [comment]
> @GitTom-HH: Are you using `GNOME Software` or `pkcon update` in Ubuntu 23.10 ?
> 
> Can you list the steps to reproduce this issue ?

I used the update-manager, this gui for the apt package management system. Just start it and wait. Components were downloaded and processing started and at some point it stopped, no more output in the log window and the process pk-debconf-helper consumed 25% CPU. You could continue processing by terminating this process. This may happen several times but in the end the packages are installed.

**Comment by sidt4** [comment]
FWIW, I ran `update-manager` on Ubuntu 23.10. It had 1 month worth of updates. Once the update was in progress at some point, `pk-debconf-helper` came in the `top` command displaying around 100% CPU. It exited after a minute or so. Here are the logs, which confirm the timing.

```
$ journalctl --user -u  pk-debconf-helper.service
Apr 01 09:14:48 ubuntu2310 systemd[1163]: Started pk-debconf-helper.service - debconf communication service.
Apr 01 09:15:48 ubuntu2310 pk-debconf-help[3218]: No active connections, exiting
Apr 01 09:15:48 ubuntu2310 systemd[1163]: pk-debconf-helper.service: Consumed 57.322s CPU time.
Apr 01 09:20:27 ubuntu2310 systemd[1163]: Started pk-debconf-helper.service - debconf communication service.
Apr 01 09:21:27 ubuntu2310 pk-debconf-help[5680]: No active connections, exiting
Apr 01 09:21:27 ubuntu2310 systemd[1163]: pk-debconf-helper.service: Consumed 47.326s CPU time.
```

All this while ( from `update-manager` starting update to finish ), here is the `pkmon` command output:

```
sid@ubuntu2310:~$ pkmon
Transactions:
 [none]
daemon connected=0
daemon connected=1
network status=online
^C
```

**Comment by GitTom-HH** [comment]
Yes, what can I say, as things stand today, I can confirm that. The service works again as it should. The last time I had the problems I described was last week when I commented on the issue and for months before that. Since then, three updates without any problems. I'll keep my fingers crossed.

**Comment by GitTom-HH** [comment]
Ok, today I saw it again. There was a Wine update and I had four or five hanging instances of the pk-debconf-helper process. After each hard termination  the next process came up and wanted also to be hard terminated. I let the first one run for about half an hour hoping it would come back on its own. But no...

This is the output of `journalctl --user -u  pk-debconf-helper.service`:

```
  -- Boot abeb415a57c4431bad8672c3764b3aa9 --
Apr 06 22:06:33 uENVY systemd[2251]: Started pk-debconf-helper.service - debconf communication service.
Apr 06 22:38:07 uENVY systemd[2251]: pk-debconf-helper.service: Consumed 31min 31.579s CPU time.
Apr 06 22:38:07 uENVY systemd[2251]: Started pk-debconf-helper.service - debconf communication service.
Apr 06 22:38:33 uENVY systemd[2251]: pk-debconf-helper.service: Consumed 25.416s CPU time.
Apr 06 22:38:41 uENVY systemd[2251]: Started pk-debconf-helper.service - debconf communication service.
Apr 06 22:39:14 uENVY systemd[2251]: pk-debconf-helper.service: Consumed 32.237s CPU time.
Apr 06 22:39:14 uENVY systemd[2251]: Started pk-debconf-helper.service - debconf communication service.
Apr 06 22:39:31 uENVY systemd[2251]: pk-debconf-helper.service: Consumed 16.707s CPU time.
Apr 06 22:39:31 uENVY systemd[2251]: Started pk-debconf-helper.service - debconf communication service.
Apr 06 22:39:53 uENVY systemd[2251]: pk-debconf-helper.service: Consumed 21.406s CPU time.

```
Afterwards  `pkmon` returns:

```
tom@uENVY:~$ pkmon
Transactions:
 [none]
daemon connected=0
daemon connected=1
network status=online
^C
```

**Comment by bennetyee** [comment]
ubuntu 24.04, 100% cpu for ~1 minute.

journalctl says

> -- Boot 57741a866c254ea5bbfa423d69aecef8 --
> Nov 28 12:10:31 lambda systemd[1351826]: Started pk-debconf-helper.service - debconf communication service.
> Nov 28 12:11:31 lambda pk-debconf-help[3383188]: No active connections, exiting
> Nov 28 12:11:31 lambda systemd[1351826]: pk-debconf-helper.service: Consumed 59.932s CPU time.

**Comment by J-Sek** [comment]
Is there any way to limit this process to.. let's say 20% of single core power? Fans on the laptop get triggered for no reason an unfortunately I cannot control their behavior.

**Comment by freonheat** [comment]
March 2025, I am seeing this behavior in Mint 22 still.  Updated a package when that was done, one core was pegged at 100% for maybe a minute.  I must have just never noticed it before now (since I have 8 cores/16 threads and was watching top for something else at the time).  This is a very fast desktop system (Ryzen 7 3700X) so I can't imagine why it could possibly need that much processing power (around a minute at 4.4Ghz is crazy).

**Comment by haroldl** [comment]
Just based on a quick code search, it does not seem like pk-debconf-helper is setting up non-blocking I/O. My guess is that even though there is nothing to do the pk_client_helper_accept_connection_cb callback is in a busy loop waiting for a new connection to come in via g_socket_accept.

Below all of the places that g_socket_set_blocking is used to enable non-blocking I/O are in files named example/test/dummy. Perhaps it is needed in pk-debconf-helper.c's main function as well.

```
$ ack g_socket_set_blocking
lib/packagekit-glib2/pk-socket-example.c
75:	g_socket_set_blocking (socket, FALSE);

lib/packagekit-glib2/pk-test-daemon.c
260:	g_socket_set_blocking (socket, FALSE);

backends/dummy/pk-backend-dummy.c
1276:		g_socket_set_blocking (job_data->socket, FALSE);
```

**Comment by johns-codings** [comment]
Linux Mint 22.3 Zena - Still 100% CPU after running update:
```shell
# journalctl --user -u pk-debconf-helper.service

-- Boot 2a77b179218d4906b2db644b5a2675e3 --
Apr 04 12:49:47 YogaPro9i systemd[4116]: Started pk-debconf-helper.service - debconf communication service.
Apr 04 12:50:48 YogaPro9i pk-debconf-help[6493]: No active connections, exiting
Apr 04 12:50:48 YogaPro9i systemd[4116]: pk-debconf-helper.service: Consumed 1min 477ms CPU time.
Apr 04 12:51:14 YogaPro9i systemd[4116]: Started pk-debconf-helper.service - debconf communication service.
Apr 04 12:52:15 YogaPro9i pk-debconf-help[9497]: No active connections, exiting
Apr 04 12:52:15 YogaPro9i systemd[4116]: pk-debconf-helper.service: Consumed 1min 475ms CPU time.
```

**Comment by freonheat** [comment]
> Linux Mint 22.3 Zena - Still 100% CPU after running update:

Yep.  I just noticed it yesterday after performing updates.  Also Mint 22.3.  I was looking at a CPU graph and thought something was wrong, then remembered this issue.

**Comment by michaelneverwins** [comment]
I also saw this on Linux Mint 22.3, ~~and there doesn't seem to be any interest in fixing it~~, so I decided to try "fixing" it myself by limiting the CPU usage of the ` pk-debconf-help` process using `cpulimit`.

I created a file `/etc/systemd/system/PackageKit-issue-452-mitigation.service` containing
```
[Unit]
Description=Uses cpulimit on pk-debconf-helper
After=multi-user.target

[Service]
ExecStart=/usr/bin/cpulimit -e pk-debconf-helper -l 1
Type=simple

[Install]
WantedBy=multi-user.target
```
and then ran the following:
```bash
sudo systemctl daemon-reload
sudo systemctl enable --now PackageKit-issue-452-mitigation
```
I suppose any other method of running `cpulimit` as `root` would work just as well, but I had the idea of creating a systemd service because I initially thought it should be triggered by the `pk-debconf-helper` service starting. Then I realized `cpulimit` can just run continuously and wait for the process, and doing it this way was easier to implement.

Anyway, it seems to have been somewhat successful. After I kicked off the next system update, I ran `pidstat -u 2 10` and saw that `pk-debconf-helper` was only using about 10% CPU, whereas previously (when I first noticed the issue) it was reporting 100%.

Meanwhile, `journalctl` output piped to `grep pk-debconf-helper` showed
```
Apr 13 21:58:37 [...] systemd[12561]: Started pk-debconf-helper.service - debconf communication service.
Apr 13 21:59:38 [...] systemd[12561]: pk-debconf-helper.service: Consumed 6.713s CPU time.
```
where the recorded CPU time is (very roughly) 10% of what it was before:
```
Apr 12 00:45:19 [...] systemd[2009]: Started pk-debconf-helper.service - debconf communication service.
Apr 12 00:46:20 [...] systemd[2009]: pk-debconf-helper.service: Consumed 1min 845ms CPU time.
```

Note that I had _expected_ only 1% CPU usage, based on the `-l` argument to `cpulimit` on the `ExecStart` line above, but I think 10% might just be the lower bound on what `cpulimit` can actually achieve, at least on this machine, for some reason. If I run `cpulimit -l 1 -f -- python busy.py`, where `busy.py` just does `1+1` in an infinite loop, another terminal running `pidstat -u ...` again reports about 10% CPU usage. Meanwhile, when I change `-l 1` to `-l 10` to run this "busy" script, `pidstat` reports it using at least 12% CPU, so maybe `cpulimit` throttling to a given percentage is just less accurate on the lower end.

But, accuracy of CPU percentages aside, throttling `pk-debconf-helper` in this way seems to work in general.

Obviously try this at your own risk. I don't _really_ know what I'm doing, besides denying excessive CPU usage to a busy-waiting process that I _assume_ will never actually need more than I'm allowing it to have. I am certainly not an expert in systemd, but I think what I've done is probably harmless and can be undone as follows:
```bash
sudo systemctl disable --now PackageKit-issue-452-mitigation
sudo rm /etc/systemd/system/PackageKit-issue-452-mitigation.service
sudo systemctl daemon-reload
```

**Comment by ximion** [comment]
I have never managed to reproduce this...
How do you call PackageKit on Mint? Does it have some special update tool? Because I have tested this with the command-line utility, GNOME Software and KDE Discover, and newer saw this issue. So potentially it is triggered by the frontend somehow, given that the original report was on Pop!_OS...

**Comment by michaelneverwins** [comment]
I see the `pk-debconf-helper` process run after installing updates through Linux Mint's "Update Manager" GUI. (This is on Cinnamon Edition; I don't know whether Xfce Edition and MATE Edition use the same program or do things differently.)

Apologies for the snark in the first sentence of my previous comment, by the way. I know what it's like to be unable to reproduce a bug. To answer this much earlier question which I didn't see until now:

>In all of these cases, does the daemon still exit after a minute?

In my case, yes, `pk-debconf-helper` does run for one minute and then stop.

**Comment by ximion** [comment]
Very strange, I took a brief look at the tool, and it doesn't seem to use PackageKit (but a really strange combination of all kinds of other stuff at the same time). So, something else must be doing this...

If this issue is reproducible, please run `pkgcli monitor` next time *before* you execute updates, then after the debconf helper has quit, exit the monitor command and paste the output here (if your PackageKit is too old for `monitor`, please run `pkmon` instead).

**Comment by mastezont** [comment]
> If this issue is reproducible, please run `pkgcli monitor`
```
$ pkgcli monitor
Command 'pkgcli' not found. You might have meant:
command 'pgcli' from the pgcli (4.0.1-1) DEB package
Try: sudo apt install <DEB package name>
```
Linux Mint Cinnamon 22.2

**Comment by ximion** [comment]
You might have `pkmon` instead then, on older distributions.

**Comment by mastezont** [comment]
> ou might have `pkmon` instead then, on older distributions.

pkmon is available. As soon as the next update is released (which could be today), I'll try using pkmon and post its output here. Do I need to specify any special parameters, or is the standard pkmon output sufficient?

**Comment by mastezont** [comment]
Updates have arrived.
`$ pkmon -v
14:47:00	PackageKit          Verbose debugging enabled (on console 1)
14:47:00	PackageKit          refreshing task list
Transactions:
 [none]
14:47:00	PackageKit          notify::connected
daemon connected=1
14:47:00	PackageKit          refreshing task list
network status=online
Daemon state: 'State:
'
14:49:03	PackageKit          emit updates-changed
updates-changed
14:49:04	PackageKit          notify::connected
daemon connected=0
14:49:04	PackageKit          notify::connected
daemon connected=1
14:49:04	PackageKit          refreshing task list
14:49:08	PackageKit          emit updates-changed
updates-changed`

[https://pastebin.com/svnnVtLt](https://pastebin.com/svnnVtLt)

<img width="691" height="196" alt="Image" src="https://github.com/user-attachments/assets/fff59b01-5a45-430b-8ac5-a8519cc4c659" />

**Comment by Lightning11wins** [comment]
I also just had this happen to me (on GNU Linux Mint). Has anyone in this thread found a solution, or even an explanation for why this happens?

**Comment by kennovo** [comment]
Let's gather what we know so far:
- This has been going on for well over 5 years (I'm gonna say about 10 but it's hard to remember)
- There seem to be hints that only some Desktop Environments trigger the problem
- On my systems (running plain Debian with the Mate desktop, but I also had it with Ubuntu and Mint in the past), it happens when gpk-update-viewer invokes pk-debconf-helper (not when running updates from the command-line, aptitude, synaptic,...)
- Symptom: as soon as all the updates are finished, pk-debconf-helper starts to consume 100% of one CPU core and continues doing so for _almost exactly_ 1 minute (again on systems using the Mate desktop)
- Interestingly, that **_duration is independent on the hardware, internet connection, or number of packages_** that have been updated. Ancient Athlon 7750 or Core 2 duo? 1 minute. Modern Ryzen 7 7700X or Core i7-12700? Still 1 minute. Marginal hotel wifi? 1 minute. Wired corporate network with its own uplink? Again 1 minute. Quick update of 2 small libraries? 1 minute. Debian point release with tens of packages? You guessed it, 1 minute.

That last point might be the key. I cannot think of any other culprit than an idle loop with a wall time limit, which should narrow down the search a lot. However (for the people getting irate at PackageKit for not yet having fixed this), it's not guaranteed that the offending idle loop is inside pk-debconf-helper. I know next to nothing about what's going on internally here, but it could theoretically be the calling process bombarding pk-debconf-helper with requests through some pipe before eventually timing out, or the caller giving pk-debconf-helper an unfortunate combination of parameters, and/or the idle loop might be in a library _invoked by_ pk-debconf-helper. So it seems we know _what_ to search for (an idle loop with a wall time limit), but not really _where_ to look for it (the loop and the limit could even be in different programs).

**Comment by mastezont** [comment]
It seems that the developers do not fully understand the logic of this application or do not consider it necessary to fix this bug.

**Comment by ximion** [comment]
> It seems that the developers do not fully understand the logic of this application

Well, to get a bugfix faster you can absolutely hire me to do that. Apart from that, issues that affect fewer people, do not cause severe issues and are mainly annoyances, and also are very hard to debug naturally take a lower priority for maintainers who are already working on way too many tasks at once.

@kennovo That 1 minute delay you observe is simply the timeout of the worker after it did not receive a connection for a while. Not receiving a connection should not cause this issue though.

I have some ideas for this, but I need to find the time to try them (tbh, there's a few issues which currently take priority).

**Comment by mastezont** [comment]
> Well, to get a bugfix faster you can absolutely hire me to do that.

I don't think this would be a rational way to spend my money. If the developers haven't been able to solve this problem in five years, then I have neither the financial ability nor the desire to pay them for another five years.

**Comment by ximion** [comment]
I get that you are frustrated because an issue that affects you is taking long to fix. But I do want to remind you both of our [Code of Conduct](https://github.com/PackageKit/PackageKit/blob/main/CODE_OF_CONDUCT.md) as well as the fact that you are talking to volunteers who are investing their time, money and expertise so that anyone can use this software for free.

This project does not have large corporate backing, and therefore progress can be slow at times.

**Comment by mastezont** [comment]
I apologize if anyone felt offended by my response.

> This project does not have large corporate backing, and therefore progress can be slow at times.

Have you considered the option of transferring support for this project to the corporate sector?

**Comment by freonheat** [comment]
> I'm sorry for veering off-topic, but I'm now incredibly curious what you imagine "transferring support to the corporate sector" to look like. Which corporate entity would you contact? What would you tell them? What's in it for them? How would you convince them to pick PackageKit over the 100s of other unfunded open source projects running on an average Linux system?

This is inappropriate and off-topic conversation for a bug report thread.

**Comment by kennovo** [comment]
Okay deleted. Bye-ee!

**Comment by phodd** [comment]
Unsure what help this might be, but I managed to catch pk-debconf-helper sticking a CPU core at 100% today after an update (LMDE 7 - unsure what other details are needed), and threw an `strace -p PID` at it a few seconds before whatever kills it (more below) did that.

The result had over 1000 lines like the following (and I strongly suspect there would have been thousands more had I been quicker off the mark):
```
ppoll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=8, events=POLLIN}, {fd=10, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 6, {tv_sec=0, tv_nsec=40423000}, NULL, 8) = 4 ([{fd=8, revents=POLLHUP}, {fd=10, revents=POLLHUP}, {fd=11, revents=POLLHUP}, {fd=13, revents=POLLHUP}], left {tv_sec=0, tv_nsec=40421878})
```

... with the tv_sec and tv_nsec counting *down*. This suggests that it is indeed in a busy loop waiting for something to happen and it's the process itself that's giving up, rather than something outside killing it.

I would have thought that the fix would be to add a `sleep 0.5` (or similar suitable value) to that loop, wherever it is, so that it only polls whatever it's polling, say, 120 times in the minute it's hanging around rather than thousands. That would presumably then have the effect of not sticking a CPU core at 100%.

The only problem with that is if it absolutely has to catch whatever it's trying to catch as soon as possible rather than dally for half a second.

FWIW, the last few `strace` output lines were (including the very last of the hundreds of `ppoll`s):

```
ppoll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=8, events=POLLIN}, {fd=10, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 6, {tv_sec=0, tv_nsec=40379000}, NULL, 8) = 4 ([{fd=8, revents=POLLHUP}, {fd=10, revents=POLLHUP}, {fd=11, revents=POLLHUP}, {fd=13, revents=POLLHUP}], left {tv_sec=0, tv_nsec=40377858}
getpeername(2, {sa_family=AF_UNIX, sun_path="/run/systemd/journal/stdout"}, [128 => 30]) = 0
futex(0x7f9326b82ed8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0) = 7
sendmsg(7, {msg_name={sa_family=AF_UNIX, sun_path="/run/systemd/journal/socket"}, msg_namelen=29, msg_iov=[{iov_base="GLIB_OLD_LOG_API", iov_len=16}, {iov_base="=", iov_len=1}, {iov_base="1", iov_len=1}, {iov_base="\n", iov_len=1}, {iov_base="MESSAGE", iov_len=7}, {iov_base="=", iov_len=1}, {iov_base="No active connections, exiting", iov_len=30}, {iov_base="\n", iov_len=1}, {iov_base="PRIORITY", iov_len=8}, {iov_base="=", iov_len=1}, {iov_base="5", iov_len=1}, {iov_base="\n", iov_len=1}], msg_iovlen=12, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 69
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
futex(0x561f323a0ad0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
close(3)                                = 0
close(5)                                = 0
kill(2713452, SIGQUIT)                  = -1 ESRCH (No such process)
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
close(6)                                = 0
kill(2713459, SIGQUIT)                  = -1 ESRCH (No such process)
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
write(4, "\1\0\0\0\0\0\0\0", 8)         = 8
sendmsg(7, {msg_name={sa_family=AF_UNIX, sun_path="/run/systemd/journal/socket"}, msg_namelen=29, msg_iov=[{iov_base="GLIB_OLD_LOG_API", iov_len=16}, {iov_base="=", iov_len=1}, {iov_base="1", iov_len=1}, {iov_base="\n", iov_len=1}, {iov_base="MESSAGE", iov_len=7}, {iov_base="=", iov_len=1}, {iov_base="g_object_unref: assertion 'G_IS_"..., iov_len=55}, {iov_base="\n", iov_len=1}, {iov_base="PRIORITY", iov_len=8}, {iov_base="=", iov_len=1}, {iov_base="4", iov_len=1}, {iov_base="\n", iov_len=1}, {iov_base="GLIB_DOMAIN", iov_len=11}, {iov_base="=", iov_len=1}, {iov_base="GLib-GObject", iov_len=12}, {iov_base="\n", iov_len=1}], msg_iovlen=16, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 119
sendmsg(7, {msg_name={sa_family=AF_UNIX, sun_path="/run/systemd/journal/socket"}, msg_namelen=29, msg_iov=[{iov_base="GLIB_OLD_LOG_API", iov_len=16}, {iov_base="=", iov_len=1}, {iov_base="1", iov_len=1}, {iov_base="\n", iov_len=1}, {iov_base="MESSAGE", iov_len=7}, {iov_base="=", iov_len=1}, {iov_base="g_object_unref: assertion 'G_IS_"..., iov_len=55}, {iov_base="\n", iov_len=1}, {iov_base="PRIORITY", iov_len=8}, {iov_base="=", iov_len=1}, {iov_base="4", iov_len=1}, {iov_base="\n", iov_len=1}, {iov_base="GLIB_DOMAIN", iov_len=11}, {iov_base="=", iov_len=1}, {iov_base="GLib-GObject", iov_len=12}, {iov_base="\n", iov_len=1}], msg_iovlen=16, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 119
exit_group(0)
```

**Comment by Lightning11wins** [comment]
That's some interesting info! Thanks for the help.

**Comment by ThomasReinhardt** [comment]
I've had the exact same issue today and stumbled over this issue. I created a reproducer by downgrading "tar" with `sudo apt install --allow-downgrades tar=1.35+dfsg-3build1` and after that upgraded again using the GUI (I'm on Linux Mint 22.3 with xfce).
Thanks to @phodd I knew where to look and took a full strace of pk-debconf-helper during that minute of CPU thrashing. With the help of claude (sorry, but read on !) I debugged the massive 550MB dump and we found this:

1,892,018 poll() calls in 60 s, of which 1,891,590 return POLLHUP, but only 297 reads on those fds in the entire minute!

Cause (rephrased and shortened):
1. the helper spawns debconf-communicate and watches its pipe/socket
2. watches are created in lib/packagekit-glib2/pk-client-helper.c like this: `source = g_io_create_watch (channel, G_IO_IN);`
3. debconf-communicate finishes and exits -> those read-ends go permanent POLLHUP
4. GLib always polls for POLLHUP/POLLERR, so poll() returns immediately
5. GIOUnixWatch's check only tests revents & watch->condition
6. watch->condition is just G_IO_IN -> POLLHUP without POLLIN -> check says "not ready" -> the callback is never dispatched -> the cleanup in pk_client_helper_copy_stdout_cb/echo_stderr_cb/copy_conn_cb never runs -> the source is never removed
7. poll() wakes instantly, dispatches nothing, loops
8. after 60 seconds the watchdog in pk-debconf-helper.c quits the whole thing

The fix is to also dispatch on POLLHUP and POLLERR to the cleanup actually can run (6+7 above).

Those changes fix the issue for me. I will create a PR for this in the next days when I have time to trim it down properly.

**Comment by ferrum** [comment]
Mint 22.3 Cinnamon, out of the box has the same issue.

**Comment by ximion** [comment]
I was thinking in the same direction of @ThomasReinhardt, who created an excellent first patch that I tweaked a little in a subsequent commit and wrote a test for to test this scenario without running debconf.

This issue should be solved now, if you can, you can compile PackageKit from source to test it. Otherwise, wait for the next release and for your distributor to provide updated packages - this will be fixed in PackageKit 1.3.7.
The issue also was introduced by accident in 2018 and does not trigger in all cases (and even if it does, many people would not notice it because the helper times out quickly). So, it's both rather old and not the easiest to notice.

Either way, thanks for sticking with it, please give feedback if you are running 1.3.7 on whether the change works well for you!
